extensions/wikia/CategoryPagination/CategoryPaginationViewer.class.php

```diff
@@ -2,102 +2,221 @@
+	 */
+	const TYPE_PAGE = 'page';
+	const TYPE_SUBCAT = 'subcat';
+	const TYPE_FILE = 'file';
+
+	/**
+	 * @var string[] content of each of the sections (excluding header and pagination)
+	 */
+	private $content = [];
+
+	/**
+	 * @var int[] number of elements in each of the sections
+	 */
+	private $counts = [];
+
+	/**
+	 * @var int[] number of elements shown on current page in each of the sections
+	 */
+	private $numberShown = [];
+
+	/**
+	 * @var int page number (from ?page URL param), always at least 1
+	 */
+	private $page = 1;
+
+	/**
+	 * @var Paginator[] paginator object for each of the sections
+	 */
 	private $paginators = [];
 
-	public function __construct( $title, IContextSource $context, $page ) {
-		parent::__construct( $title, $context );
+	/**
+	 * @var CategoryTree
+	 */
+	private $tree = null;
+
+	/**
+	 * Get the category page HTML
+	 *
+	 * @return string
+	 */
+	public function getHTML() {
+		global $wgCategoryMagicGallery, $wgEnableCategoryTreeExt, $wgCategoryTreeCategoryPageOptions;
 
-		$this->page = $page;
+		// Image galleries
+		$this->showGallery = $wgCategoryMagicGallery && !$this->getOutput()->mNoGallery;
+		$this->clearCategoryState();
+
+		// Initialize category tree
+		if ( !empty( $wgEnableCategoryTreeExt ) && !$this->getRequest()->getCheck( 'notree' ) ) {
+			$this->tree = new CategoryTree( $wgCategoryTreeCategoryPageOptions );
+			$this->tree->setHeaders( $this->getOutput() );
+		}
+
+		$this->generateData();
+
+		// Hooks for Answers and CategoryGalleries
+		/** @see answerCategoryOtherSection */
+		$answersSection = '';
+		wfRunHooks( 'CategoryViewer::getOtherSection', [ &$this, &$answersSection ] );
+		/** @see CategoryGalleriesHelper::onCategoryPageGetCategoryTop */
+		$categoryGallery = '';
+		wfRunHooks( 'CategoryPage::getCategoryTop', [ $this, &$categoryGallery ] );
+
+		// Render
+		$this->getOutput()->addHeadItem( 'Paginator', $this->getPaginationHeadItem() );
+		$tpl = new EasyTemplate( __DIR__ . '/templates' );
+		$tpl->set_vars( [
+			'aswersSection' => $answersSection,
+			'categoryGallery' => $categoryGallery,
+			'content' => $this->content,
+			'counts' => $this->counts,
+			'empty' => empty( $answersSection ) && empty( $this->content ),
+			'lang' => $this->getLanguage(),
+			'numberShown' => $this->numberShown,
+			'paginator' => $this->paginators,
+			'titleEscaped' => htmlspecialchars( $this->getTitle()->getText() ),
+		] );
+		return $tpl->render( 'category' );
+	}
 
-		$cat = Category::newFromTitle( $this->title );
+	/**
+	 * Generate the data for each of the default sections
+	 * Save it to $this->content and $this->numberShown
+	 */
+	private function generateData() {
+		// Get counts
+		$cat = $this->getCat();
 		$this->counts = [
-			'file' => $cat->getFileCount(),
-			'subcat' => $cat->getSubcatCount(),
-			'page' => $cat->getPageCount() - $cat->getFileCount() - $cat->getSubcatCount(),
+			self::TYPE_FILE => $cat->getFileCount(),
+			self::TYPE_SUBCAT => $cat->getSubcatCount(),
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use "static" keyword instead of "self". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2037)